### ISSUE TITLE
Model Performance Engineering

### ISSUE THREAD

**Issue by nirmalpratheep** [The-School-of-AI/LLM#222]
## 🎯 Objective (Performance Only)

Design, measure, and lock in the **fastest repeatable training configuration** for the **1B dense seed model** with:
- ✅ Provable GPU efficiency (kernel-level validation)
- ✅ Maximum memory bandwidth utilization
- ✅ Stable end-to-end throughput (tokens/sec)

**Scope**: CUDA execution, kernel efficiency, memory layout, scheduling, compute/communication overlap, profiling, and metrics.

**Not in Scope**: Model architecture, training logic, hyperparameter tuning (correctness only).

---

### Training Configuration
```bash
# Baseline command
python scripts/train_pipeline.py \
    --hidden_size 2048 \
    --num_layers 24 \
    --batch_size 4 \
    --learning_rate 6e-4 \
    --output_dir outputs/1b_dense
```

**Model Specs**:
- **Parameters**: ~1B (dense model)
- **Hidden size**: 2048
- **Layers**: 24
- **Batch size**: 4 (per GPU)
- **Sequence length**: 2048 (default)

**Hardware**:
- GPU: A100 80GB (or specify)
- GPU count: 1-8 (scaling tests)
- Interconnect: NVLink / PCIe

---

### 1️⃣ Performance Benchmark Report

**File**: `performance_report_1b_dense.md`

**Required Metrics**:

| Category | Metric | Unit | Target |
|----------|--------|------|--------|
| **Throughput** | Tokens/sec (global, steady-state) | tok/s | Max achievable |
| | Samples/sec | samples/s | - |
| | Wall-clock time per step | ms | <500ms |
| **Latency** | Step time p50 | ms | - |
| | Step time p95 | ms | <p50 * 1.1 |
| | Step time p99 | ms | <p50 * 1.2 |
| **GPU Efficiency** | GPU utilization (SM) | % | >85% |
| | Memory bandwidth utilization | % | >70% |
| | Kernel efficiency score | % | >80% |
| **Stability** | Throughput variance (std/mean) | % | <3% |
| | Step time jitter | ms | <20ms |
| **Theoretical** | Max theoretical throughput | tok/s | - |
| | Achieved % of theoretical | % | >75% |

### 2️⃣ Kernel-Level Analysis

**File**: `kernel_analysis.md`

**Required Sections**:
- [ ] **Top 10 kernels by time** (name, duration, % of total)
- [ ] **Bottleneck attribution** (compute-bound vs memory-bound)
- [ ] **Kernel launch overhead** analysis
- [ ] **Memory access patterns** (coalesced vs uncoalesced)
- [ ] **Occupancy analysis** (warps, registers, shared memory)
- [ ] **Recommendations** for kernel optimization

### 3️⃣ Memory Profile

**File**: `memory_profile.md`

**Required Data**:
- [ ] Peak memory usage per GPU (MB)
- [ ] Memory bandwidth utilization timeline
- [ ] Memory allocation breakdown (activations, weights, gradients, optimizer)
- [ ] Fragmentation analysis
- [ ] Memory access patterns (read/write ratio)
- [ ] Opportunities for memory reuse

### 4️⃣ System Timeline

**File**: `system_timeline.html` (Nsight Systems output)

**Must Show**:
- [ ] CUDA kernel execution timeline
- [ ] CPU thread activity
- [ ] NCCL communication (if multi-GPU)
- [ ] Data loading pipeline
- [ ] Gaps/stalls identification
- [ ] Compute-communication overlap visualization

### Phase 1: System-Level Timeline (Nsight Systems)

**Objective**: Identify high-level bottlenecks (gaps, stalls, overlap issues)

```bash
# Full system timeline
nsys profile \
  --trace=cuda,nvtx,osrt,cudnn,cublas,nccl \
  --cuda-memory-usage=true \
  --gpu-metrics-device=all \
  --capture-range=cudaProfilerApi \
  --capture-range-end=stop \
  --stats=true \
  -o seed_1b_timeline \
  python scripts/train_pipeline.py \
    --hidden_size 2048 \
    --num_layers 24 \
    --batch_size 4 \
    --learning_rate 6e-4 \
    --output_dir outputs/1b_dense \
    --profile_steps 10-20
```

**Output**: `seed_1b_timeline.nsys-rep`

**Analysis Steps**:
1. Open in Nsight Systems GUI
2. Identify gaps in GPU utilization
3. Check CPU-GPU synchronization points
4. Measure kernel launch overhead
5. Verify compute-communication overlap (multi-GPU)
6. Export summary stats: `nsys stats seed_1b_timeline.nsys-rep`

**Deliverables**:
- [ ] Screenshot of timeline (annotated with bottlenecks)
- [ ] Summary statistics (markdown table)
- [ ] List of top 10 time-consuming operations

### Phase 2: Kernel-Level Analysis (Nsight Compute)

**Objective**: Deep-dive into individual kernel performance

```bash
# Profile all kernels (WARNING: slow, for analysis only)
ncu \
  --set full \
  --target-processes all \
  --import-source yes \
  --section SpeedOfLight \
  --section MemoryWorkloadAnalysis \
  --section ComputeWorkloadAnalysis \
  --section Occupancy \
  -o seed_1b_kernels \
  python scripts/train_pipeline.py \
    --hidden_size 2048 \
    --num_layers 24 \
    --batch_size 4 \
    --learning_rate 6e-4 \
    --output_dir outputs/1b_dense \
    --profile_steps 10-11  # Profile single step
```

**Output**: `seed_1b_kernels.ncu-rep`

**Alternative (Focused Profiling)**:
```bash
# Profile only top kernels by name (faster)
ncu \
  --set full \
  --kernel-name-base demangled \
  --kernel-regex "attention|matmul|layernorm|gelu" \
  -o seed_1b_focused_kernels \
  python scripts/train_pipeline.py ...
```

**Analysis Steps**:
1. Open in Nsight Compute GUI
2. Sort kernels by duration
3. For each top kernel:
   - Check **Speed of Light** (SOL) metrics
   - Identify limiting factor (compute vs memory)
   - Check **Occupancy** (theoretical vs achieved)
   - Analyze **Memory Workload** (coalescing, cache hit rates)
   - Review **Compute Workload** (warp stall reasons)
4. Export metrics: `ncu --import seed_1b_kernels.ncu-rep --csv > kernel_metrics.csv`

**Deliverables**:
- [ ] Top 10 kernels table (name, time, %, bottleneck)
- [ ] Detailed analysis for top 3 kernels
- [ ] Kernel optimization recommendations

### Phase 3: Memory Profiling (PyTorch Profiler)

**Objective**: Understand memory allocation and bandwidth usage

```python
# Add to train_pipeline.py
import torch
from torch.profiler import profile, ProfilerActivity, tensorboard_trace_handler

with profile(
    activities=[ProfilerActivity.CPU, ProfilerActivity.CUDA],
    schedule=torch.profiler.schedule(wait=5, warmup=5, active=10, repeat=1),
    on_trace_ready=tensorboard_trace_handler('./profiler_logs'),
    record_shapes=True,
    profile_memory=True,
    with_stack=True
) as prof:
    for step, batch in enumerate(dataloader):
        # Training step
        outputs = model(batch)
        loss = outputs.loss
        loss.backward()
        optimizer.step()
        optimizer.zero_grad()
        
        prof.step()
        
        if step >= 20:
            break

# Print summary
print(prof.key_averages().table(sort_by="cuda_time_total", row_limit=20))
print(prof.key_averages().table(sort_by="cuda_memory_usage", row_limit=20))

# Export
prof.export_chrome_trace("memory_profile.json")
```

**Alternative (Memory Snapshot)**:
```python
import torch.cuda._memory_viz as memory_viz

# During training
torch.cuda.memory._record_memory_history(max_entries=100000)

# After training
snapshot = torch.cuda.memory._snapshot()
with open("memory_snapshot.pickle", "wb") as f:
    pickle.dump(snapshot, f)

# Visualize: python -m torch.cuda._memory_viz trace_plot memory_snapshot.pickle
```

**Deliverables**:
- [ ] Memory allocation timeline
- [ ] Peak memory breakdown (pie chart)
- [ ] Memory bandwidth utilization over time
- [ ] Opportunities for memory reduction

### Phase 4: Throughput Scaling Analysis

**Objective**: Measure throughput scaling with batch size and GPU count

```bash
# Batch size scaling (single GPU)
for bs in 1 2 4 8 16; do
    python scripts/train_pipeline.py \
        --hidden_size 2048 \
        --num_layers 24 \
        --batch_size $bs \
        --profile_steps 10-110 \
        --output_dir outputs/bs_${bs} \
        2>&1 | tee logs/bs_${bs}.log
done

# GPU scaling (fixed batch size)
for ngpu in 1 2 4 8; do
    torchrun --nproc_per_node=$ngpu \
        scripts/train_pipeline.py \
        --hidden_size 2048 \
        --num_layers 24 \
        --batch_size 4 \
        --profile_steps 10-110 \
        --output_dir outputs/gpu_${ngpu} \
        2>&1 | tee logs/gpu_${ngpu}.log
done
```

**Analysis**:
```python
# parse_scaling.py
import re
import matplotlib.pyplot as plt

# Parse logs for tokens/sec
def parse_throughput(log_file):
    with open(log_file) as f:
        for line in f:
            if "tokens/sec" in line:
                match = re.search(r"(\d+\.?\d*)\s*tokens/sec", line)
                if match:
                    return float(match.group(1))
    return None

# Batch size scaling
batch_sizes = [1, 2, 4, 8, 16]
throughputs_bs = [parse_throughput(f"logs/bs_{bs}.log") for bs in batch_sizes]

# GPU scaling
gpu_counts = [1, 2, 4, 8]
throughputs_gpu = [parse_throughput(f"logs/gpu_{n}.log") for n in gpu_counts]

# Plot
fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(12, 5))

ax1.plot(batch_sizes, throughputs_bs, marker='o')
ax1.set_xlabel("Batch Size")
ax1.set_ylabel("Tokens/sec")
ax1.set_title("Batch Size Scaling")
ax1.grid(True)

ax2.plot(gpu_counts, throughputs_gpu, marker='o', label='Actual')
ax2.plot(gpu_counts, [throughputs_gpu[0] * n for n in gpu_counts], 'r--', label='Ideal Linear')
ax2.set_xlabel("GPU Count")
ax2.set_ylabel("Tokens/sec")
ax2.set_title("GPU Scaling Efficiency")
ax2.legend()
ax2.grid(True)

plt.tight_layout()
plt.savefig("scaling_analysis.png", dpi=300)
```

**Deliverables**:
- [ ] Batch size scaling plot
- [ ] GPU scaling efficiency plot
- [ ] Recommended batch size per GPU
- [ ] Scaling efficiency table

---

### 1. Setup Environment
```bash
# Install profiling tools
pip install torch torchvision torchaudio
pip install nvidia-pyindex
pip install nvidia-dlprof[pytorch]

# Install Nsight tools (download from NVIDIA)
# - Nsight Systems 2024.x
# - Nsight Compute 2024.x

# Verify
nsys --version
ncu --version
```

### 2. Run Baseline
```bash
# Single GPU, no profiling
python scripts/train_pipeline.py \
    --hidden_size 2048 \
    --num_layers 24 \
    --batch_size 4 \
    --learning_rate 6e-4 \
    --output_dir outputs/baseline \
    --profile_steps 10-110

# Capture baseline tokens/sec
```

### 3. Run Nsight Systems
```bash
nsys profile -o baseline_timeline \
    python scripts/train_pipeline.py \
    --hidden_size 2048 \
    --num_layers 24 \
    --batch_size 4 \
    --profile_steps 10-20

# Open: nsys-ui baseline_timeline.nsys-rep
```

### 4. Analyze & Iterate
- Identify top bottleneck
- Test optimization
- Re-profile
- Compare metrics
- Repeat until targets met

---

## 📝 Example Output (Performance Report)

```markdown
# Performance Report: 1B Dense Model

**Date**: 2026-01-31
**Hardware**: 4x A100 80GB, NVLink
**Configuration**: See `optimal_config_1b_dense.json`

## Summary

✅ **All targets met** - Configuration locked and ready for production

## Key Metrics

| Metric | Value | Target | Status |
|--------|-------|--------|--------|
| Tokens/sec (1 GPU) | 12,345 | ≥10,000 | ✅ |
| Tokens/sec (4 GPU) | 46,123 | - | - |
| GPU Utilization | 88.3% | ≥85% | ✅ |
| Memory Bandwidth | 74.1% | ≥70% | ✅ |
| Throughput Variance | 2.3% | <3% | ✅ |
| Step Time (p50) | 412 ms | - | - |
| Step Time (p95) | 438 ms | <453 ms | ✅ |

## Bottleneck Analysis

**Top Kernel**: `flash_attn_fwd` (32.1% of step time)
- Compute-bound (SM utilization: 91.2%)
- Occupancy: 87.3% (theoretical: 100%)
- **Recommendation**: Already optimal (Flash Attention 2)

**#2 Kernel**: `gemm_bias_act` (18.7% of step time)
- Memory-bound (bandwidth util: 68.4%)
- **Recommendation**: Enable `torch.compile()` for fusion

**Communication**: All-reduce overlap efficiency: 94.1% ✅

## Optimizations Applied

1. ✅ Flash Attention 2 → 2.3x speedup vs standard
2. ✅ Fused LayerNorm → 1.2x speedup
3. ✅ FusedAdam → 1.15x speedup
4. ✅ `torch.compile(mode="max-autotune")` → 1.4x overall
5. ✅ bf16 mixed precision → 1.8x vs fp32

**Total Speedup**: 5.2x vs baseline naive implementation

## Recommendations for Team 9

- Use `batch_size=4` per GPU (optimal for A100 80GB)
- Enable gradient checkpointing if memory tight on ZeRO-3
- Monitor all-reduce time: should be <15% of step time
- Expect 90-95% scaling efficiency on 4 GPUs

## Configuration Lock

See `optimal_config_1b_dense.json` for full details.

**Validated on**: 2026-01-31  
**Performance valid through**: Model architecture changes only
```

---

**Comment by nirmalpratheep** [comment]
This was added to the flow. Added nsight system,pytorch profiler and compute to the  flow.
